Task 4

evTyurin opened the pull request:
Complete task 4

YevganyTyurin committed:
Update .gitignore

YevganyTyurin committed:
Complete task 4

YevganyTyurin committed:
Add task description

YevganyTyurin committed:
Add class diagram

YevganyTyurin committed:
Complete task 4.1

YevganyTyurin committed:
Complete task 4.2

mrseneca reviewed (changes requested):
Proper git-flow needed.
We agreed to merge 3.4 task into some branch what will be development branch for future.
But your [Electronic_car_service_administrator](https://github.com/evTyurin/SENLA-java-courses/tree/Electronic_car_service_administrator) branch is empty
You can look up Mark's comment about this topic in our skype chat

evTyurin commented:
My bad. I moved my commits from [Task 3.4](https://github.com/evTyurin/SENLA-java-courses/tree/TASK_3.4) to [Electronic_car_service_administrator](https://github.com/evTyurin/SENLA-java-courses/tree/Electronic_car_service_administrator) previously but didn't push it. Now I fixed it.

mrseneca commented on src/com/senlainc/warsaw/tyurin/entity/Garage.java:
```diff
@@ -0,0 +1,77 @@
+import java.util.List;
+import java.util.Objects;
+
+public class Garage {
```
Why exactly you need Entity, Repo and Service for Garage? I thought your FULL application should represent some Garage. But right now looks like you have a bunch of garages, with bunch of mechanics. Lets not do this, and change for Garage representation.

mrseneca replied:
So this entity, repo for it and also service should be deleted.

mrseneca commented on src/com/senlainc/warsaw/tyurin/entity/OrderStatus.java:
```diff
@@ -0,0 +1,9 @@
+package com.senlainc.warsaw.tyurin.entity;
+
+public enum OrderStatus {
```
It's not Entity, so you can move it to different package: util or variable for example

mrseneca commented on src/com/senlainc/warsaw/tyurin/entity/Craftsman.java:
```diff
@@ -0,0 +1,76 @@
+    private long id;
+    private String name;
+    private String surname;
+    private Map<LocalDateTime, Boolean> schedule;
```
No, scheduling should not be part of the mechanic. "Schedule" will be an abstraction that is based on orders. In the order, you must have the date of its execution, the id of the mechanic and the id of the place in the garage. And based on this, on the go it is necessary to calculate when and who is free on a certain date. There is no need to store it, especially in the entity mechanic

mrseneca commented on src/com/senlainc/warsaw/tyurin/entity/GaragePlace.java:
```diff
@@ -0,0 +1,77 @@
+    private long id;
+    private int number;
+    private double space;
+    private Map<LocalDateTime, Boolean> schedule;
```
Same about schedule as for mechanic

mrseneca commented on src/com/senlainc/warsaw/tyurin/service/GaragePlaceService.java:
```diff
@@ -0,0 +1,130 @@
+
+    private static GaragePlaceService INSTANCE;
+    private IGaragePlaceDAO garagePlaceDAO;
+    private ICraftsmanDAO craftsmanDAO;
```
Unfortunately, it is bad practice to use someone else's dao in an inappropriate service. If you really need to work with the mechanics in this service, you can do it through the mechanics service, not through the dao.

mrseneca replied:
But think about pass desired mechanic to method what need it, if it's possible.

mrseneca commented on src/com/senlainc/warsaw/tyurin/Test.java:
```diff
@@ -0,0 +1,208 @@
+        CraftsmanService craftsmanService = CraftsmanService.getInstance();
+        OrderService orderService = OrderService.getInstance();
+
+        Craftsman bob = new Craftsman();
```
Mechanics, orders and garage places should be created on Service layer, not here (you can count it as Controller layer).
So pass needed variables to service methods and in there create entities and save them to dao

mrseneca commented on src/com/senlainc/warsaw/tyurin/Test.java:
```diff
@@ -0,0 +1,208 @@
+        order3.addCraftsman(ted);
+        order3.addCraftsman(bob);
+
+        GaragePlaceDAO garagePlaceDAO = GaragePlaceDAO.getInstance();
```
You can not operate with dao on this layer, here you can operate only with Service.
So as i wrote in comment before, create and save entities through service methods

mrseneca reviewed (changes requested):
Good job, but you need to fix the architecture to make it easier for you to develop in the future and use best practices

YevganyTyurin committed:
Correct task for PR comments

mrseneca reviewed (approved):
That will do work for now. Still some small issues in code and architecture, but wil will fix them later. Good job

evTyurin commented:
> That will do work for now. Still some small issues in code and architecture, but wil will fix them later. Good job

Thank you :)
